File: low_storage_2.rb
=begin
	For this one, I used insertion sort to sort the array. It's very low storage [O(1)],
because the only data structure that it uses is the passed in array. Time-wise,
though, it's inefficient.

  To find the missing number in the sorted array,  I used a method that also iterates 
through the passed in array without generating any additional data structures, making
it low storage as well.

To use this code call low_storage_2(array_with_missing_val) 

Ex:
hundy = gen_array(80)
missing = low_storage_2(hundy)
puts "missing: #{missing}"
=end

# sorts by moving elements into the correct position of a sorted portion of the array
def insertion_sort(array)
	for i in 1...array.length
		j = i
		# keep going as long as the array isn't sorted
		while j > 0 && array[j-1] > array[j]
			# when the previous element is larger, swap positions with it
			array[j-1], array[j] = array[j], array[j-1]
			j = j - 1
		end
	end
	array
end

# iterates through the array, comparing adjacent elements until the missing number is found.
def find_missing(array)
	# edge cases
	return 1 if array[0] != 1
	return array.length + 1 if array[array.length - 1] != array.length + 1
	array.each_with_index do |el, i|
		# if next element is greater by more than 1, missing number is it minus 1.
		return array[i] + 1 if array[i+1] - array[i] > 1
	end
	# when the array is not missing a number
	return -1;
end

def gen_array(missing)
	array = []
	(1..100).each do |i|
		array << i unless i == missing
	end
	array.shuffle
end

def low_storage_2(array)
	sorted = insertion_sort(array)
	find_missing(sorted)
end

hundy = gen_array(10)
missing = low_storage_2(hundy)
puts "missing: #{missing}"

File: low_storage_2.js
/*
I chose to sort via bubble sort, because it doesn't create any data structures and 
will be very low storage. This makes it O(1). The function that I use to then find the 
missing values will be low storage for the same reason.

To use it, call the function lowStorageTwo(arrayWithMissingVal)

Ex:
var hunds = genArray(3);
var valMissing = lowStorageTwo(hunds);
console.log('valMissing', valMissing);

*/

function lowStorageTwo(array){
	bubbleSort(array);
	return findMissing(array);
}

var hunds = genArray(3);
var valMissing = lowStorageTwo(hunds);
console.log('valMissing', valMissing);

// iterates through the array, comparing adjacent elements until the missing number is found.
function findMissing(array){
	// edge cases
	if(array[0] != 1) return 1;
	if(array[array.length - 1] != array.length + 1) return array.length + 1;
	for(var i = 1; i < array.length; i++){
			// if next element is greater by more than 1, missing number is it minus 1.
			if(array[i] - array[i-1] > 1) return array[i-1] + 1;
	}
	// when the array is not missing a number
	return -1;
}

// Compares adjacent values, swapping higher ones for lower ones until the array is sorted. 
function bubbleSort(array){
	var swapped = true,
			temp;
	while(swapped){
		swapped = false;
		for(var i = 1; i < array.length; i++){
			if(array[i-1] > array[i]){
				// make the swap
				temp = array[i-1];
				array[i-1] = array[i];
				array[i] = temp;
				swapped = true;
			}
		}
	}
}

// Generates a randomly sorted (for the most part) array. The argument is the missing number.
function genArray(missing){
	var hundo = [];
	for(var i = 1; i <= 100; i++){
		if(i != missing) hundo.push(i);
	}
	hundo.sort(function(){
		return Math.round(Math.random()) - 0.5;
	});
	return hundo
}